packages/graphback-core/src/plugin/GraphbackPlugin.ts

```diff
@@ -0,0 +1,51 @@
+import { GraphQLObjectType, GraphQLSchema } from "graphql";
+import { parseMarker } from 'graphql-metadata';
+import { getUserTypesFromSchema } from './getUserTypesFromSchema';
+
+/**
+ * Graphback plugin interface
+ * Plugins are base for every graphback generator and schema transformers. 
+ * See documentation for the complete list of the plugins.
+ * 
+ * Plugins can:
+ * 
+ * - Modify the schema
+ * - Create resources like files, db tables etc.
+ * - Perform some in memory operations based on configuration
+ */
+export abstract class GraphbackPlugin {
+
+    /**
+     * Performs transformation on the schema and returns target schemaƒ
+     * @param schema 
+     */
+    public abstract transformSchema(schema: GraphQLSchema): GraphQLSchema;
+
+    /**
+     * @returns Unique name of the plugin
+     */
+    public abstract getPluginName(): string;
+
+    protected logWarning(message: string): void {
+        // tslint:disable-next-line: no-console
+        console.log(`Warning - ${this.getPluginName()}: ${message}`)
+    }
+
+    protected logError(message: string): void {
+        // tslint:disable-next-line: no-console
+        console.error(`Error - ${this.getPluginName()}: ${message}`)
+    }
+
+    /**
+     * Helper for plugins to fetch all types that should be processed by Graphback plugins.
+     * To mark type as enabled for graphback generators we need to add `model` annotations over the type. 
+     * 
+     * Returns all user types that have @model in description
+     * @param schema 
+     */
+    protected getUserModels(schema: GraphQLSchema): GraphQLObjectType[] {
```

Comment:
> Would you see a value in this function be moved into a separate file not in this class as a standalone helper function?

If we move to helper files then this refactoring is pointless and we do not have architecture done right. The whole purpose is to have things working with each other and core that drives some information forward.

Reply:
Having this function in the base `GraphbackPlugin` implies types marked with `@model` will be needed in most cases. But the idea of the plugin architecture is that it is generic enough to fit all use cases.

Reply:
> Helper for plugins to fetch all types that should be processed by Graphback plugins

What if I want a plugin that does some transformation of non-model type? This JSDoc strongly implies that the plugin engine does not cater for this.

Reply:
Actually I would need to think about it. It was originally as a helper. 
We need some kind of core for the entire ecosystem. 

So we can have 2 layers in the core. Generator plugin and general ModelFacade that gives information for the schema. 

So I would see it that plugin by default will have this facade created and whatever else will need it will simply need to create new opbject for a schema